File: src/main/java/de/bypixeltv/redivelocity/managers/RedisController.java
package de.bypixeltv.redivelocity.managers;

import de.bypixeltv.redivelocity.RediVelocity;
import de.bypixeltv.redivelocity.config.Config;
import lombok.Getter;
import redis.clients.jedis.BinaryJedisPubSub;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPoolConfig;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import org.json.JSONObject;

import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.atomic.AtomicBoolean;

@Singleton
public class RedisController extends BinaryJedisPubSub implements Runnable {

    private final RediVelocity rediVelocity;
    @Getter
    private final JedisPool jedisPool;
    private final byte[][] channelsInByte;
    private final AtomicBoolean isConnectionBroken = new AtomicBoolean(true);
    private final AtomicBoolean isConnecting = new AtomicBoolean(false);

    @Inject
    public RedisController(RediVelocity rediVelocity, Config config) {
        this.rediVelocity = rediVelocity;

        JedisPoolConfig jConfig = new JedisPoolConfig();
        int maxConnections = 10;

        jConfig.setMaxTotal(maxConnections);
        jConfig.setMaxIdle(maxConnections);
        jConfig.setMinIdle(1);
        jConfig.setBlockWhenExhausted(true);

        String password = config.getRedis().getPassword();
        if (password.isEmpty()) {
            this.jedisPool = new JedisPool(jConfig, config.getRedis().getHost(), config.getRedis().getPort());
        } else {
            this.jedisPool = new JedisPool(jConfig, config.getRedis().getHost(), config.getRedis().getPort(), 2000, password);
        }

        this.channelsInByte = setupChannels();
    }

    @Override
    public void run() {
        if (!isConnectionBroken.get() || isConnecting.get()) {
            return;
        }
        rediVelocity.sendLogs("Connecting to Redis server...");
        isConnecting.set(true);
        try (var jedis = jedisPool.getResource()) {
            isConnectionBroken.set(false);
            rediVelocity.sendLogs("Connection to Redis server has established! Success!");
            jedis.subscribe(this, channelsInByte);
        } catch (Exception e) {
            isConnecting.set(false);
            isConnectionBroken.set(true);
            rediVelocity.sendErrorLogs("Connection to Redis server has failed! Please check your details in the configuration.");
            e.printStackTrace();
        }
    }

    public void shutdown() {
        jedisPool.close();
    }

    public void sendPostLoginMessage(String event, String proxyId, String username, String useruuid, String userip, String channel) {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("action", event);
        jsonObject.put("proxy", proxyId);
        jsonObject.put("name", username);
        jsonObject.put("uuid", useruuid);
        jsonObject.put("address", userip);
        jsonObject.put("timestamp", System.currentTimeMillis());
        String jsonString = jsonObject.toString();

        try (var jedis = jedisPool.getResource()) {
            jedis.publish(channel, jsonString);
        }
    }

    public void sendServerSwitchMessage(String event, String proxyId, String username, String useruuid, String clientbrand, String userip, String serverName, String previousServer, String channel) {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("action", event);
        jsonObject.put("proxy", proxyId);
        jsonObject.put("name", username);
        jsonObject.put("uuid", useruuid);
        jsonObject.put("clientbrand", clientbrand);
        jsonObject.put("address", userip);
        jsonObject.put("timestamp", System.currentTimeMillis());
        jsonObject.put("server", serverName);
        jsonObject.put("previousServer", previousServer);

        String jsonString = jsonObject.toString();

        try (var jedis = jedisPool.getResource()) {
            jedis.publish(channel, jsonString);
        }
    }

    public void sendJsonMessage(String event, String proxyId, String username, String useruuid, String clientbrand, String userip, String channel) {
        JSONObject jsonObject = new JSONObject();
        jsonObject.put("action", event);
        jsonObject.put("proxy", proxyId);
        jsonObject.put("name", username);
        jsonObject.put("uuid", useruuid);
        jsonObject.put("clientbrand", clientbrand);
        jsonObject.put("address", userip);
        jsonObject.put("timestamp", System.currentTimeMillis());
        String jsonString = jsonObject.toString();

        try (var jedis = jedisPool.getResource()) {
            jedis.publish(channel, jsonString);
        }
    }

    public void sendMessage(String message, String channel) {
        try (var jedis = jedisPool.getResource()) {
            jedis.publish(channel, message);
        }
    }

    public void removeFromListByValue(String listName, String value) {
        try (var jedis = jedisPool.getResource()) {
            jedis.lrem(listName, 0, value);
        }
    }

    public void setHashField(String hashName, String fieldName, String value) {
        try (var jedis = jedisPool.getResource()) {
            String type = jedis.type(hashName);
            if (!"hash".equals(type)) {
                if ("none".equals(type)) {
                    jedis.hset(hashName, fieldName, value);
                } else {
                    System.err.println("Error: Key " + hashName + " doesn't hold a hash. It holds a " + type + ".");
                }
            } else {
                jedis.hset(hashName, fieldName, value);
            }
        }
    }

    public void deleteHashField(String hashName, String fieldName) {
        try (var jedis = jedisPool.getResource()) {
            jedis.hdel(hashName, fieldName);
        }
    }

    public void deleteHash(String hashName) {
        try (var jedis = jedisPool.getResource()) {
            jedis.del(hashName);
        }
    }

    public void addToList(String listName, String[] values) {
        try (var jedis = jedisPool.getResource()) {
            for (String value : values) {
                jedis.rpush(listName, value);
            }
        }
    }

    public void setListValue(String listName, int index, String value) {
        try (var jedis = jedisPool.getResource()) {
            long listLength = jedis.llen(listName);
            if (index >= listLength) {
                System.err.println("Error: Index " + index + " does not exist in the list " + listName + ".");
            } else {
                jedis.lset(listName, index, value);
            }
        }
    }

    public Map<String, String> getHashValuesAsPair(String hashName) {
        Map<String, String> values = new HashMap<>();
        try (var jedis = jedisPool.getResource()) {
            Set<String> keys = jedis.hkeys(hashName);
            for (String key : keys) {
                values.put(key, jedis.hget(hashName, key));
            }
        }
        return values;
    }

    public void removeFromList(String listName, int index) {
        try (var jedis = jedisPool.getResource()) {
            long listLength = jedis.llen(listName);
            if (index >= listLength) {
                System.err.println("Error: Index " + index + " does not exist in the list " + listName + ".");
            } else {
                String tempKey = UUID.randomUUID().toString();
                jedis.lset(listName, index, tempKey);
                jedis.lrem(listName, 0, tempKey);
            }
        }
    }

    public void deleteList(String listName) {
        try (var jedis = jedisPool.getResource()) {
            jedis.del(listName);
        }
    }

    public void setString(String key, String value) {
        try (var jedis = jedisPool.getResource()) {
            jedis.set(key, value);
        }
    }

    public String getString(String key) {
        try (var jedis = jedisPool.getResource()) {
            return jedis.get(key);
        }
    }

    public void deleteString(String key) {
        try (var jedis = jedisPool.getResource()) {
            jedis.del(key);
        }
    }

    public String getHashField(String hashName, String fieldName) {
        try (var jedis = jedisPool.getResource()) {
            return jedis.hget(hashName, fieldName);
        }
    }

    public Set<String> getAllHashFields(String hashName) {
        try (var jedis = jedisPool.getResource()) {
            return jedis.hkeys(hashName);
        }
    }

    public List<String> getAllHashValues(String hashName) {
        try (var jedis = jedisPool.getResource()) {
            return jedis.hvals(hashName);
        }
    }

    public List<String> getList(String listName) {
        try (var jedis = jedisPool.getResource()) {
            return jedis.lrange(listName, 0, -1);
        }
    }

    public String getHashValueByField(String hashName, String fieldName) {
        try (var jedis = jedisPool.getResource()) {
            return jedis.hget(hashName, fieldName);
        }
    }

    public Boolean exists(String key) {
        try (var jedis = jedisPool.getResource()) {
            return jedis.exists(key);
        }
    }

    public List<String> getHashFieldNamesByValue(String hashName, String value) {
        List<String> fieldNames = new ArrayList<>();
        try (var jedis = jedisPool.getResource()) {
            Set<String> keys = jedis.keys(hashName);
            for (String key : keys) {
                Map<String, String> fieldsAndValues = jedis.hgetAll(key);
                for (Map.Entry<String, String> entry : fieldsAndValues.entrySet()) {
                    if (entry.getValue().equals(value)) {
                        fieldNames.add(entry.getKey());
                    }
                }
            }
        }
        return fieldNames;
    }

    private byte[][] setupChannels() {
        List<String> channels = Arrays.asList("global", "messaging", "friends", "utils", "other"); // replace with your actual channels
        byte[][] channelsInByte = new byte[channels.size()][];
        for (int i = 0; i < channels.size(); i++) {
            channelsInByte[i] = channels.get(i).getBytes(StandardCharsets.UTF_8);
        }
        return channelsInByte;
    }

}

File: src/main/kotlin/de/bypixeltv/redivelocity/managers/RedisController.kt
package de.bypixeltv.redivelocity.managers

import de.bypixeltv.redivelocity.RediVelocity
import com.velocitypowered.api.scheduler.ScheduledTask
import de.bypixeltv.redivelocity.config.Config
import org.json.JSONObject
import redis.clients.jedis.BinaryJedisPubSub
import redis.clients.jedis.JedisPool
import redis.clients.jedis.JedisPoolConfig
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

class RedisController(private val plugin: RediVelocity, private val config: Config) : BinaryJedisPubSub(), Runnable {

    private val jedisPool: JedisPool
    private var channelsInByte: Array<ByteArray>
    private val isConnectionBroken = AtomicBoolean(true)
    private val isConnecting = AtomicBoolean(false)
    private var connectionTask: ScheduledTask? = null

    init {
        val jConfig = JedisPoolConfig()
        val maxConnections = 10

        jConfig.maxTotal = maxConnections
        jConfig.maxIdle = maxConnections
        jConfig.minIdle = 1
        jConfig.blockWhenExhausted = true

        val password = config.redisPassword ?: ""
        jedisPool = if (password.isEmpty()) {
            JedisPool(
                jConfig,
                config.redisHost,
                config.redisPort,
                9000,
                config.useSsl
            )
        } else {
            JedisPool(
                jConfig,
                config.redisHost ?: "0.0.0.0",
                config.redisPort,
                9000,
                password,
                config.useSsl
            )
        }

        channelsInByte = setupChannels()
        connectionTask = plugin.proxy.scheduler.buildTask(plugin, this).repeat(5, TimeUnit.SECONDS).schedule()
    }

    override fun run() {
        if (!isConnectionBroken.get() || isConnecting.get()) {
            return
        }
        plugin.sendLogs("Connecting to Redis server...")
        isConnecting.set(true)
        try {
            jedisPool.resource.use { jedis ->
                isConnectionBroken.set(false)
                plugin.sendLogs("Connection to Redis server has established! Success!")
                jedis.subscribe(this, *channelsInByte)
            }
        } catch (e: Exception) {
            isConnecting.set(false)
            isConnectionBroken.set(true)
            plugin.sendErrorLogs("Connection to Redis server has failed! Please check your details in the configuration.")
            e.printStackTrace()
        }
    }

    fun shutdown() {
        connectionTask?.cancel()
        if (this.isSubscribed) {
            try {
                this.unsubscribe()
            } catch (e: Exception) {
                plugin.sendErrorLogs("Something went wrong during unsubscribing...")
                e.printStackTrace()
            }
        }
        jedisPool.close()
    }

    fun sendJsonMessage(event: String, proxyId: String, username: String, useruuid: String, clientbrand: String, userip: String, ping: Int, channel: String) {
        val jsonObject = JSONObject()
        jsonObject.put("action", event)
        jsonObject.put("proxyid", proxyId)
        jsonObject.put("username", username)
        jsonObject.put("uuid", useruuid)
        jsonObject.put("clientbrand", clientbrand)
        jsonObject.put("ipadress", userip)
        jsonObject.put("ping", ping)
        jsonObject.put("timestamp", System.currentTimeMillis())

        val jsonString = jsonObject.toString()

        // Publish the JSON string to the specified channel
        jedisPool.resource.use { jedis ->
            jedis.publish(channel, jsonString)
        }
    }

    fun sendMessage(message: String, channel: String) {
        jedisPool.resource.use { jedis ->
            jedis.publish(channel, message)
        }
    }

    fun removeFromListByValue(listName: String, value: String) {
        jedisPool.resource.use { jedis ->
            jedis.lrem(listName, 0, value)
        }
    }

    fun setHashField(hashName: String, fieldName: String, value: String) {
        jedisPool.resource.use { jedis ->
            val type = jedis.type(hashName)
            if (type != "hash") {
                if (type == "none") {
                    jedis.hset(hashName, fieldName, value)
                } else {
                    System.err.println("Error: Key $hashName doesn't hold a hash. It holds a $type.")
                }
            } else {
                jedis.hset(hashName, fieldName, value)
            }
        }
    }

    fun deleteHashField(hashName: String, fieldName: String) {
        jedisPool.resource.use { jedis ->
            jedis.hdel(hashName, fieldName)
        }
    }

    fun deleteHash(hashName: String) {
        jedisPool.resource.use { jedis ->
            jedis.del(hashName)
        }
    }

    fun addToList(listName: String, values: Array<String>) {
        jedisPool.resource.use { jedis ->
            values.forEach { value ->
                jedis.rpush(listName, value)
            }
        }
    }

    fun setListValue(listName: String, index: Int, value: String) {
        jedisPool.resource.use { jedis ->
            val listLength = jedis.llen(listName)
            if (index >= listLength) {
                System.err.println("Error: Index $index does not exist in the list $listName.")
            } else {
                jedis.lset(listName, index.toLong(), value)
            }
        }
    }

    fun getHashValuesAsPair(hashName: String): Map<String, String> {
        val values = mutableMapOf<String, String>()
        jedisPool.resource.use { jedis ->
            val keys = jedis.hkeys(hashName)
            for (key in keys) {
                values[key] = jedis.hget(hashName, key)
            }
        }
        return values
    }

    fun removeFromList(listName: String, index: Int) {
        jedisPool.resource.use { jedis ->
            val listLength = jedis.llen(listName)
            if (index >= listLength) {
                System.err.println("Error: Index $index does not exist in the list $listName.")
            } else {
                val tempKey = UUID.randomUUID().toString()
                jedis.lset(listName, index.toLong(), tempKey)
                jedis.lrem(listName, 0, tempKey)
            }
        }
    }

    fun deleteList(listName: String) {
        jedisPool.resource.use { jedis ->
            jedis.del(listName)
        }
    }

    fun setString(key: String, value: String) {
        jedisPool.resource.use { jedis ->
            jedis.set(key, value)
        }
    }

    fun getString(key: String): String? {
        return jedisPool.resource.use { jedis ->
            jedis.get(key)
        }
    }

    fun deleteString(key: String) {
        jedisPool.resource.use { jedis ->
            jedis.del(key)
        }
    }

    fun getHashField(hashName: String, fieldName: String): String? {
        return jedisPool.resource.use { jedis ->
            jedis.hget(hashName, fieldName)
        }
    }

    fun getAllHashFields(hashName: String): Set<String>? {
        return jedisPool.resource.use { jedis ->
            jedis.hkeys(hashName)
        }
    }

    fun getAllHashValues(hashName: String): List<String>? {
        return jedisPool.resource.use { jedis ->
            jedis.hvals(hashName)
        }
    }

    fun getList(listName: String): List<String>? {
        return jedisPool.resource.use { jedis ->
            jedis.lrange(listName, 0, -1)
        }
    }

    fun getHashFieldNamesByValue(hashName: String, value: String): List<String> {
        val fieldNames = mutableListOf<String>()
        jedisPool.resource.use { jedis ->
            val keys = jedis.keys(hashName)
            for (key in keys) {
                val fieldsAndValues = jedis.hgetAll(key)
                for (entry in fieldsAndValues.entries) {
                    if (entry.value == value) {
                        fieldNames.add(entry.key)
                    }
                }
            }
        }
        return fieldNames
    }


    private fun setupChannels(): Array<ByteArray> {
        val channels = listOf("global", "messaging", "friends", "utils", "other") // replace with your actual channels
        return Array(channels.size) { channels[it].toByteArray(StandardCharsets.UTF_8) }
    }

    fun isRedisConnectionOffline(): Boolean {
        return isConnectionBroken.get()
    }

    fun getJedisPool(): JedisPool {
        return jedisPool
    } //

}